cmd/auth/auth.go

```diff
@@ -0,0 +1,200 @@
+			return
+		}
+
+		jsonData, err := json.Marshal(sessionData)
+		if err != nil {
+			http.Error(w, "Error marshaling session data", http.StatusInternalServerError)
+			return
+		}
+
+		filePath := path.Join(dirName, sessionFileName)
+		if err := os.WriteFile(filePath, jsonData, 0o600); err != nil {
+			http.Error(w, "Error writing to file", http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "text/html; charset=utf-8")
+		_, err = w.Write([]byte(successHTML))
+		if err != nil {
+			fmt.Println("Error writing response:", err)
+			return
+		}
+
+		username := sessionData["user"].(map[string]interface{})["user_metadata"].(map[string]interface{})["user_name"].(string)
+		fmt.Println("🎉 Login successful 🎉")
+		fmt.Println("Welcome aboard", username, "🍕")
+
+	})
+
+	err = browser.OpenURL(authenticationURL)
+	if err != nil {
+		fmt.Println("Failed to open the browser 🤦‍♂️")
+		fmt.Println("Navigate to the following URL to begin authentication:")
+		fmt.Println(authenticationURL)
+	}
+
+	errCh := make(chan error)
+	go func() {
+		errCh <- server.ListenAndServe()
+	}()
+
+	interruptCh := make(chan os.Signal, 1)
+	signal.Notify(interruptCh, os.Interrupt)
+
+	select {
+	case err := <-errCh:
+		if err != nil && err != http.ErrServerClosed {
+			return err
+		}
+	case <-time.After(60 * time.Second):
+		shutdown(server)
+		return errors.New("authentication timeout")
+	case <-interruptCh:
+		fmt.Println("\nAuthentication interrupted❗️")
+		shutdown(server)
+		os.Exit(0)
+	}
+	return nil
+}
+
+func getSession(authCode, codeVerifier string) (map[string]interface{}, error) {
+	url := fmt.Sprintf("https://%s.supabase.co/auth/v1/token?grant_type=pkce", supabaseID)
+
+	payload := map[string]string{
+		"auth_code":     authCode,
+		"code_verifier": codeVerifier,
+	}
+
+	jsonPayload, _ := json.Marshal(payload)
+
+	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(jsonPayload))
+	req.Header.Set("Content-Type", "application/json;charset=UTF-8")
+	req.Header.Set("ApiKey", supabasePublicKey)
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status: %s", res.Status)
+	}
+
+	var responseData map[string]interface{}
+	if err := json.NewDecoder(res.Body).Decode(&responseData); err != nil {
+		return nil, err
+	}
+
+	return responseData, nil
+}
+
+func pkce(length int) (verifier, challenge string, err error) {
+	p := make([]byte, length)
+	if _, err := io.ReadFull(rand.Reader, p); err != nil {
+		return "", "", err
+	}
+	verifier = base64.RawURLEncoding.EncodeToString(p)
+	b := sha256.Sum256([]byte(verifier))
+	challenge = base64.RawURLEncoding.EncodeToString(b[:])
+	return verifier, challenge, nil
+}
+
+func shutdown(server *http.Server) {
+	go func() {
+		err := server.Shutdown(context.Background())
+		if err != nil {
+			fmt.Println("Graceful shutdown failed", err)
```

Comment:
This error message could be abit better. And may make sense to use `panic` here since the graceful shutdown should almost never happen.